File: LeetCode/Easy/0234-palindrome-linked-list/0234-palindrome-linked-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public bool IsPalindrome(ListNode head) {

        var slow = head;
        var fast = head;

        while (fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }

        var reversed = ReverseList(slow);
        var areListsEqual = AreListEqual(head, reversed);

        ReverseList(reversed);
        return areListsEqual;
    }

    private ListNode ReverseList(ListNode node)
    {
        ListNode previous = null;
        ListNode current = node;

        while (current != null)
        {
            var next = current.next;
            current.next = previous;
            previous = current;
            current = next;
        }

        return previous;
    }

    private bool AreListEqual(ListNode first, ListNode second)
    {
        while (second != null)
        {
            if (first.val != second.val) return false;

            first = first.next;
            second = second.next;
        }

        return true;
    }
}

File: LeetCode/Easy/0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        def find_after_mid_node(current: ListNode) -> ListNode:
            slow = head
            fast = head
            while fast and fast.next:
                fast = fast.next.next
                slow = slow.next
            
            return slow

        def reverse_list(current: ListNode) -> ListNode:
            previous = None
            while current:
                current.next, previous, current = previous, current, current.next 
            
            return previous

        def compare_reversed_with_straight(reversed_half: ListNode, straight_half: ListNode) -> bool:
            second_half_head = reversed_half
            while reversed_half:
                if straight_half.val != reversed_half.val:
                    reverse_list(second_half_head)
                    return False
                reversed_half = reversed_half.next
                straight_half = straight_half.next
        
            reverse_list(second_half_head)
            return True

        second_half_head = reverse_list(find_after_mid_node(head))
        return compare_reversed_with_straight(second_half_head, head)
        
        

        